windows/src/engine/keyman32/calldll.cpp

```diff
@@ -253,48 +347,215 @@ void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum)
+  //SendDebugMessageFormat(0, sdmKeyboard, 0, "KMSetOutput: Enter");
   PKEYMAN64THREADDATA _td = ThreadGlobals();
-  if(!_td) return FALSE;
-	if(!_td->app) return FALSE;
-        SendDebugMessageFormat(0, sdmKeyboard, 0, "KMQueueAction: Exit");
-	return _td->app->QueueAction(ItemType, dwData);
+  if (!_td)
+    return FALSE;
+  if (!_td->app)
+    return FALSE;
+
+  if (!Globals::get_CoreIntegration()) {  // TODO: 5442 Remove If and fix indent
+    while (backlen-- > 0)
+      _td->app->QueueAction(QIT_BACK, 0);
+    while (*buf)
+      _td->app->QueueAction(QIT_CHAR, *buf++);
+    SendDebugMessageFormat(0, sdmKeyboard, 0, "KMSetOutput: Exit");
+    return TRUE;
+  } else {
+    if (!_td->lpActiveKeyboard->lpCoreKeyboardState) {
+      SendDebugMessageFormat(0, sdmKeyboard, 0, "KMSetOutputCore: no active state");
+      return FALSE;
+    }
+    DWORD numActions = backlen + (DWORD)wcslen(buf);
+    DWORD idx = 0;
+    km_kbp_action_item *actionItems = new km_kbp_action_item[numActions + 1];
+
+    // The actions sent to the core processor need to set the expected_type
+    // correctly. To do this need to check the context as we process the
+    // backspaces.
+    km_kbp_context_item *citems = nullptr;
+    if (KM_KBP_STATUS_OK !=
+        (km_kbp_status_codes)kbp_state_get_intermediate_context(_td->lpActiveKeyboard->lpCoreKeyboardState, &citems)) {
+      km_kbp_context_items_dispose(citems);
+      return FALSE;
+    }
+    WCHAR *contextString           = new WCHAR[MAXCONTEXT];  // *3 if every context item was a deadkey
+    if (!ContextItemToAppContext(citems, contextString, MAXCONTEXT)) {
+      km_kbp_context_items_dispose(citems);
+      delete[] contextString;
+      return FALSE;
+    }
+    km_kbp_context_items_dispose(citems);
+    AppContext context;
+    context.Set(contextString);
+    delete[] contextString;
+    SendDebugMessageFormat(0, sdmKeyboard, 0, "KMSetOutputCore: backlen[%lu],buflen[%lu]", backlen, (DWORD)wcslen(buf));
+    while (backlen-- > 0) {
+      actionItems[idx].type = KM_KBP_IT_BACK;
+      WCHAR *CodeUnitPtr;
+      const int DeadKeyLength = 3;
+      const int SurrogateLength = 2;
+      const int SingleCharLength = 1;
+      if (context.CharIsDeadkey()) {
+        CodeUnitPtr = context.BufMax(DeadKeyLength);
+        CodeUnitPtr += 2;
+        actionItems[idx].backspace.expected_type  = KM_KBP_BT_MARKER;
+        actionItems[idx].backspace.expected_value = (uintptr_t)*CodeUnitPtr;
+      } else if (context.CharIsSurrogatePair()) {
+        CodeUnitPtr = context.BufMax(SurrogateLength);
+        actionItems[idx].backspace.expected_type  = KM_KBP_BT_CHAR;
+        actionItems[idx].backspace.expected_value = (DWORD)Uni_SurrogateToUTF32(*CodeUnitPtr, *(CodeUnitPtr + 1));
+      } else {
+        CodeUnitPtr = context.BufMax(SingleCharLength);
+        actionItems[idx].backspace.expected_type  = KM_KBP_BT_CHAR;
+        actionItems[idx].backspace.expected_value = (DWORD)*CodeUnitPtr;
```

Comment:
What about the case where we are deleting off the start of the context (i.e. context is empty)?